README
==================
SILVA <VERSION> README
==================

SilVA: Silent variant analysis using random forests.
"When a synonymous mutation falls in the exome, does it make a sound?"


SilVA is a tool for the automated harmfulness prediction of SYNONYMOUS single-nucleotide variants. Given variants in a VCF file, SilVA will rank the rare synonymous variants according to their predicted harmfulness. SilVA predicts the harmfulness of mutations using features that include conservation, codon usage, splice sites, splicing enhancers and suppressors, and mRNA folding free energy.

If you use this tool, please cite:

    Buske OJ, Manickaraj A, Mital S, Ray PN, Brudno M. (2013)
    Identification of deleterious synonymous variants in human genomes.
    Bioinformatics, doi:10.1093/bioinformatics/btt308

Send questions, comments, and difficulties to: silva-snv@cs.toronto.edu


Quickstart
==========

1. Download SilVA:

    wget http://compbio.cs.toronto.edu/silva/release/silva-<VERSION>.tar.gz
    tar -xzf silva-<VERSION>.tar.gz
    cd silva-<VERSION>

2. Install dependencies:

    ./setup.sh

3. Preprocess VCF file:

    ./silva-preprocess OUTDIR VCF

4. Run models and print highest-scoring synonymous variants:

    ./silva-run OUTDIR | head


Overview
========

Prerequisites
-------------
- Linux or Mac OS, x86
- 6GB of RAM
- Python 2.6 or 2.7 in your PATH
-- The 'numpy' Python package
- Perl in your PATH (for maxentscan)
- R in your PATH (for randomForest)

Dependencies
------------
SilVA requires several tools and databases to run. Most of these were included with this release, but the 'setup.sh' script will download and configure the rest.

- included: maxentscan, available from:
  http://genes.mit.edu/burgelab/maxent/download/
- included: twobitreader, available on pypi:
  https://pypi.python.org/pypi/twobitreader
- setup.sh: randomForest package, available on CRAN:
  http://cran.r-project.org/web/packages/randomForest/
- setup.sh: UNAfold, available from:
  http://mfold.rna.albany.edu/?q=DINAMelt/software
- setup.sh: ViennaRNA, available from:
  http://www.tbi.univie.ac.at/~ronny/RNA/
- setup.sh: SilVA databases, available from:
  http://compbio.cs.toronto.edu/silva/release/silva-<VERSION>_data.tar.gz
- setup.sh: UCSC hg19 multifasta file, available from:
  http://compbio.cs.toronto.edu/silva/release/hg19.fa.gz

* Note: if you have the SilVA databases already installed, you can point SilVA to them by changing the appropriate lines in the 'init.sh' file.

Data files
----------
After running setup.sh, all necessary data files should be automatically downloaded into the data/ directory. Here is a list of the files that should be there:
- refGene.ucsc.gz
- 1000gp.refGene.vcf.gz
- gerp.refGene.table.gz
- hg19.fa.gz

* Note: The very first time you run SilVA, it will take much longer than normal (~30min longer) because the reference genome needs to be processed and mapped to the refSeq gene annotations and data files need to be parsed. This only needs to be done once, since the following pre-processed databases are saved to the data/ directory for future runs:
- refGene.pkl
- 1000gp.refGene.pkl
- gerp.refGene.pkl

Input File Format
-----------------
Single-nucleotide variants should be given to SilVA in VCF or a VCF-like file format. This file may be gzip'd, but then must have a '.gz' extension. There should be one line per variant, with tab-delimited fields:

1. chrom - chromosome (any 'chr' prefix will be trimmed)
2. pos - the position of the variant, 1-indexed
3. id - the id field is ignored, but is here to provide compatibility with the VCF format.
4. ref - the reference nucleotide
5. alt - the alternate nucleotide (if multiple are present, comma-separated, only the first will be used)
... Additional columns can be present and are ignored

Output Format
-------------
SilVA will print the synonymous variants to stdout, ordered by score, with the variants most likely to be harmful listed first. Variants are first filtered down to just rare and novel synonymous variants. The variants are then annotated with a number of features and the machine learning model is used to rank the variants. The output contains the following tab-delimited columns:

1. The variant rank, out of all synonymous variants considered.
2. The SilVA score, between 0 and 1, where close to 1 is more likely to be harmful.
3. The variant classification, based upon several SilVA score thresholds.
3. HGNC gene name
4. RefSeq mRNA identifier
5. HGVS coding mutation identifier
... Fields from the input file


Installation
============

SilVA is packaged with most of its dependencies. The remaining few can be downloaded and configured by running the 'setup.sh' script from the root directory of this package:

1. Download, untar, and unzip the package tarball:

    wget http://compbio.cs.toronto.edu/silva/release/silva-<VERSION>.tar.gz
    tar -xzf silva-<VERSION>.tar.gz
    cd silva-<VERSION>

2. Run the setup script in the package's root directory:

    ./setup.sh

* SilVA uses a number of environment variables to communicate important paths and parameters, such as what directory to use for temporary files and what allele frequency threshold to use. SilVA has default settings that should work, but will defer to any settings in your environment (so you can hard-code values by exporting variables in your ~/.bashrc, for example). To see a list of these variables or to change their settings, see the 'init.sh' script.


Running SilVA
=============

In addition to running SilVA on your local machine, we provide a convenience script to dispatch multiple SilVA runs across an SGE cluster. SilVA uses TMPDIR for all temporary files. If TMPDIR is not set, it will use the current directory.

Local machine
-------------
To run SilVA on your local machine, you should have already installed dependencies with the 'setup.sh' script. You can then run SilVA on your file of variants, <VCF>, (see Input Format) from the root directory of this package (or add this directory to your PATH):

1. Filter an annotate the variants in the VCF file (this takes a while):

    ./silva-preprocess <OUTDIR> <VCF>

2. Run the trained models on the annotated variants and print the top variants in order of decreasing predicted harmfulness (this is fast):

    ./silva-run <OUTDIR> | head

SilVA should handle errors and early termination gracefully, so if something goes wrong, you can re-run the same command and it will pick up where it left off.

Example
-------
The 'example/' folder in the root directory of this package contains a sample input file for you to test SilVA on. You can run SilVA on it with the following commands:

    ./silva-preprocess example example/example.vcf
    ./silva-run example | head

You should see the known-harmful ACVRL1 variant appear at the top! Granted, for simplicity, the example only has 28 other rare synonymous variants in it. SilVA's expected output on this example is included in 'example/example.out'.

SGE Cluster
-----------
In case you have a large number of VCF files and an SGE cluster, we have provided a dispatch script for your convenience, 'sge/dispatch.sh'. Because SGE configurations vary widely, you should first open the file and take a look at the three SGE configuration variables at the top of the script. Edit them as you see fit. You can then dispatch SGE jobs for a number of VCF files from the root directory of the package with:

    ./sge/dispatch.sh <OUTDIR> <VCF1> <VCF2> ...


Support
=======

Send questions, comments, and difficulties to: silva-snv@cs.toronto.edu


HISTORY
2.0.0:
	- Code is now only designed to generate table of scores for tabix accessing

1.1.1: 12 July 2013
	- Switched from hg19.fa as part of data tarball to hg19.2bit from UCSC
	- Added no-folding release, without UNAfold and ViennaRNA dependencies
	- Improved some error reporting

1.1.0: 10 July 2013
	- Silently updated ViennaRNA download URL

1.1.0: 24 April 2013
	- Streamlined pipeline
	- Added vienna RNA folding features
	- Improved input and output formats slightly

1.0.1: 30 September 2012
	- Now using R randomForest package instead of Python milk package
	- Output format changed
	- Removed WEKA as a requirement/dependency
	- Added an example to the distribution

1.0.0: 13 May 2012
	- First public release.


setup.sh
#!/usr/bin/env bash


set -eu
set -o pipefail


######################################################################
# Uncomment the line below to exclude RNA folding (UNAfold/ViennaRNA)
# from the installation process:

#EXCLUDE_RNA_FOLDING=1

######################################################################

version="$(cat VERSION)"
SILVA_DATA_URL="http://compbio.cs.toronto.edu/silva/release/silva-${version}_data.tar.gz"

GENOME_URL="http://compbio.cs.toronto.edu/silva/release/hg19.fa.gz"

UNAFOLD_VERSION="3.8"
UNAFOLD_URL="http://mfold.rna.albany.edu/cgi-bin/UNAFold-download.cgi?unafold-${UNAFOLD_VERSION}.tar.gz"

VIENNA_VERSION="2.1.1"
VIENNA_URL="http://www.tbi.univie.ac.at/~ronny/RNA/packages/source/ViennaRNA-${VIENNA_VERSION}.tar.gz"

# Check python verison
pyversion=$(python -c "import sys; print sys.version[:3]")
if [[ $pyversion != 2.6 && $pyversion != 2.7 ]]; then
    cat >&2 <<EOF
Found Python $pyversion, but requires Python 2.6 or 2.7.

Please install Python 2.6/2.7, or set your PATH such that the 'python'
command loads Python 2.6 or 2.7.
EOF
    exit 1
fi

function prompt {
    echo -e "$@" >&2
    read -p "(Press ENTER to continue, Ctrl-C to exit) "
}

echo -e "\nInstalling SilVA $version dependencies..." >&2


function get_and_make_package {
    local URL=$1
    local from=$2
    local to=$3
    
    pushd tools > /dev/null
    if [[ ! -e $from.tar.gz ]]; then
	prompt "\n\nDownloading $from..."
        wget -O $from.tar.gz "$URL"
    fi
    if [[ ! -d $to ]]; then
	prompt "\n\nUnpacking $from.tar.gz to tools/$to..."
        tar -xzvf $from.tar.gz
        mv $from $to
    fi
    pushd $to > /dev/null
    prompt "\n\nConfiguring and making $to..."
    ./configure && make
    popd > /dev/null
    popd > /dev/null
}

if [ -z "${EXCLUDE_RNA_FOLDING:-}" ]; then
    # Download unafold
    if [[ ! -e tools/unafold/src/hybrid-ss-min ]]; then
	get_and_make_package $UNAFOLD_URL unafold-${UNAFOLD_VERSION} unafold
    fi

    # Make vienna
    if [[ ! -e tools/vienna/Progs/RNAfold ]]; then
	get_and_make_package $VIENNA_URL ViennaRNA-${VIENNA_VERSION} vienna
    fi
else
    echo "EXCLUDE_RNA_FOLDING=${EXCLUDE_RNA_FOLDING}... excluding RNA folding packages." >&2
fi

# Install R randomForest package
function is_rf_missing {
    echo 'suppressMessages(require("randomForest", quietly=TRUE))' \
        | R --vanilla --quiet --slave 2>&1
}

function rf_fail {
    cat <<EOF

Problem encountered trying to automatically install the randomForest module
for R. Please make sure R is installed, in your path, and you have installed
the randomForest module before running SilVA. If you start R, randomForest
can be installed by typing install.packages("randomForest").

EOF
    exit 1
}


# Install randomForest R package
if [[ "$(is_rf_missing)" ]]; then
    prompt "Installing randomForest package for R..."
    echo 'install.packages("randomForest", repos="http://probability.ca/cran/")' \
	| R --vanilla --quiet || rf_fail
fi
if [[ "$(is_rf_missing)" ]]; then
    rf_fail
fi


# Download data
datafile=data/refGene.ucsc.gz
if [[ ! -e $datafile ]]; then
    tarball=silva-${version}_data.tar.gz
    if [[ ! -e $tarball ]]; then
	prompt "\n\nDownloading required SilVA $version databases ($tarball)..."
	wget -v "$SILVA_DATA_URL"
    fi
    if [[ ! -e $datafile ]]; then
	echo -e "\n\nUnpacking required SilVA $version databases into data/..." >&2
	if [[ -e $tarball ]]; then
	    tar -xzvf $tarball
	else
	    echo -e "\n\nError: missing file: $tarball"
	    exit 1
	fi
    fi
fi

# Download hg19 multifasta file
mkdir -pv data
datafile=data/hg19.fa.gz
if [[ ! -e $datafile ]]; then
    if [[ ! -e $datafile ]]; then
	prompt "\n\nFinal step:\nDownloading hg19 FASTA file from UofT servers...\n\nThis file is large (~800M) and may take a while to download.\n\nIf you already have hg19.fa.gz, press CTRL+C to exit this script and make a symlink to it within the data directory. Then, re-run this script to make sure everything is properly linked."
	wget -v -O $datafile $GENOME_URL
    fi
    if [[ ! -e $datafile ]]; then
	echo -e "\n\nError: missing file: $datafile"
	exit 1
    fi
fi


cat >&2 <<EOF


SilVA $version successfully installed.

To run SilVA on a VCF file, first preprocess the file with 
'./silva-preprocess', and then generate results from the output 
directory with './silva-run'
EOF

